src/main/java/com/google/sps/data/Election.java

```diff
@@ -14,81 +14,166 @@
-  public static Election fromJSONObject(JSONObject obj) throws JSONException {
+  @AutoValue.Builder
+  public abstract static class Builder {
+    public abstract Builder setId(String id);
+
+    public abstract Builder setName(String name);
+
+    public abstract Builder setScope(String scope);
+
+    public abstract Builder setDate(String date);
+
+    public abstract Builder setContests(HashSet<String> contests);
+
+    public abstract Builder setPropositions(HashSet<String> propositions);
+
+    public abstract Election build();
+  }
+
+  /**
+   * Creates a new Election object by extracting the properties from "electionQueryData".
+   *
+   * @param electionQueryData the JSON output of an electionQuery to the Google Civic Information
+   *     API
+   * @return the new Election object
+   */
+  public static Election fromElectionQuery(JSONObject electionQueryData) throws JSONException {
     return Election.builder()
-        .setId(obj.getLong("id"))
-        .setName(obj.getString("name"))
-        .setDate(obj.getString("electionDay"))
-        .setScope(obj.getString("ocdDivisionId"))
-        // TODO(caseyprice): get values for Contests and Propositions
-        .setContests(new HashSet<Long>())
-        .setPropositions(new HashSet<Long>())
+        .setId(electionQueryData.getString("id"))
+        .setName(electionQueryData.getString("name"))
+        .setDate(electionQueryData.getString("electionDay"))
+        .setScope(electionQueryData.getString("ocdDivisionId"))
+        .setContests(new HashSet<String>())
+        .setPropositions(new HashSet<String>())
         .build();
   }
 
-  // creates a new Entity and sets the proper properties.
-  public Entity toEntity() {
+  /**
+   * Creates a new Election object by extracting the properties from an Election object and from the
+   * properties of "voterInfoQueryData". Delegates creating Contest Entities from this election's
+   * list of contests.
+   *
+   * @param election the Election object already populated by fromElectionQuery
+   * @param datastore the Datastore to store this election's list of contests
+   * @param voterInfoQueryData the JSON output of a voterInfoQuery to the Google Civic Information
+   *     API
+   * @return the new Election object
+   */
+  public static Election fromVoterInfoQuery(
+      Election election, DatastoreService datastore, JSONObject voterInfoQueryData)
+      throws JSONException {
+    HashSet<String> contestKeyList = new HashSet<String>();
+    HashSet<String> propositionKeyList = new HashSet<String>();
+
+    if (voterInfoQueryData.has("contests")) {
+      JSONArray contestListData = voterInfoQueryData.getJSONArray("contests");
+      for (Object contestObject : contestListData) {
+        JSONObject contest = (JSONObject) contestObject;
+
+        Contest.fromVoterInfoQuery(datastore, contest).putInDatastore(datastore);
+      }
+
+      contestKeyList = ServletUtils.getEntityKeyNameList(datastore, "Contest");
```

Comment:
Wouldn't this return the keys for *all* contests? e.g. if you had Election A and Election B, where Election A has contests A1, A2.  Election B has contests B1, B2.

I believe getEntityKeyNameList returns [A1, A2, B1, B2], where we probably only want [A1, A2] for Election A.

Maybe you can just obtain the contestKeyList from the `putInDatastore()` call on ln 102?

Reply:
You're right, I didn't catch that. I'll make that fix!